domain/machine/state/machine_cloud_instance.go

```diff
@@ -0,0 +1,153 @@
+	}
+	return row.toHardwareCharacteristics(), nil
+}
+
+// SetInstanceData sets an entry in the instance data table along with
+// the instance tags and the link to a lxd profile if any.
+func (st *State) SetInstanceData(
+	ctx context.Context,
+	machineUUID string,
+	instanceID instance.Id,
+	hardwareCharacteristics instance.HardwareCharacteristics,
+) error {
+	db, err := st.DB()
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	setInstanceData := `
+INSERT INTO machine_cloud_instance (*)
+VALUES ($instanceData.*)
+`
+	setInstanceDataStmt, err := st.Prepare(setInstanceData, instanceData{})
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	setInstanceTags := `
+INSERT INTO instance_tag (*)
+VALUES ($instanceTag.*)
+`
+	setInstanceTagStmt, err := st.Prepare(setInstanceTags, instanceTag{})
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	return db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		instanceData := instanceData{
+			MachineUUID:          machineUUID,
+			InstanceID:           string(instanceID),
+			Arch:                 hardwareCharacteristics.Arch,
+			Mem:                  hardwareCharacteristics.Mem,
+			RootDisk:             hardwareCharacteristics.RootDisk,
+			RootDiskSource:       hardwareCharacteristics.RootDiskSource,
+			CPUCores:             hardwareCharacteristics.CpuCores,
+			CPUPower:             hardwareCharacteristics.CpuPower,
+			AvailabilityZoneUUID: hardwareCharacteristics.AvailabilityZone,
+			VirtType:             hardwareCharacteristics.VirtType,
+		}
+		if err := tx.Query(ctx, setInstanceDataStmt, instanceData).Run(); err != nil {
+			return errors.Annotatef(domain.CoerceError(err), "inserting instance data for machine %q", machineUUID)
+		}
+		for _, tag := range *hardwareCharacteristics.Tags {
+			instanceTag := instanceTag{
+				MachineUUID: machineUUID,
+				Tag:         tag,
+			}
+			if err := tx.Query(ctx, setInstanceTagStmt, instanceTag).Run(); err != nil {
+				return errors.Annotatef(domain.CoerceError(err), "inserting instance tag %q for machine %q", tag, machineUUID)
+			}
+		}
+		return nil
+	})
+}
+
+// DeleteInstanceData removes an entry in the instance data table along with
+// the instance tags and the link to a lxd profile if any.
+func (st *State) DeleteInstanceData(
+	ctx context.Context,
+	machineUUID string,
+) error {
+	db, err := st.DB()
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	deleteInstanceData := `
+DELETE FROM machine_cloud_instance 
+WHERE machine_uuid=$instanceData.machine_uuid
+`
+	deleteInstanceDataStmt, err := st.Prepare(deleteInstanceData, instanceData{})
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	deleteInstanceTags := `
+DELETE FROM instance_tag 
+WHERE machine_uuid=$instanceTag.machine_uuid
+`
+	deleteInstanceTagStmt, err := st.Prepare(deleteInstanceTags, instanceTag{})
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	return db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		instanceData := instanceData{
```

Comment:
Create these at the top and use them in `Prepare`.

Reply:
Done, see 29982a633a.